libbeat/processors/add_docker_metadata/watcher.go

```diff
@@ -143,17 +163,22 @@ func (w *watcher) watch() {
 					image := event.Actor.Attributes["image"]
 					delete(event.Actor.Attributes, "name")
 					delete(event.Actor.Attributes, "image")
+
+					w.Lock()
 					w.containers[event.Actor.ID] = &Container{
 						ID:     event.Actor.ID,
 						Name:   name,
 						Image:  image,
 						Labels: event.Actor.Attributes,
 					}
+					w.Unlock()
 				}
 
 				// Delete
 				if event.Action == "die" || event.Action == "kill" {
```

Comment:
If a container dies and gets started again it will still time out eventually and lose the metadata.

Reply:
I think that case is treated, as `onPodAdd` (also called in case of update) cleans any deleted flag: https://github.com/elastic/beats/pull/5084/files#diff-c211c9481fb346b82d8713e779e0be9fR141

Reply:
Sorry you are right, I was talking about the Kubernetes watcher, not Docker, will apply the same fix, thanks!

I'm also working on tests for both watchers

Reply:
pushed https://github.com/elastic/beats/pull/5084/commits/b54187e6d907895f1bff3678338bb1e91fc5945a

Reply:
When I restart a container, the events are: kill -> die -> stop -> start -> restart. So the container will not be un-deleted in this case.

Looking at it, adding the metadata on "created" and removing on "die" does seem inconsistent to me.

Reply:
Let me do some tests, looks like it would be better to add it on start, isn't it?

Reply:
I think adding on start is the better approach.